src/arch/ucx/machine.C

```diff
@@ -323,28 +323,41 @@ static inline UcxRequest* UcxPostRxReq(ucp_tag_t tag, size_t size,
 
     // Request completed immediately
     if (req->completed) {
-        int idx = req->idx;
-
         if (!(tag & UCX_RMA_TAG_MASK)) {
             handleOneRecvedMsg(size, (char*)buf);
         }
-
-        UCX_REQUEST_FREE(req);
-
-        if (tag & UCX_MSG_TAG_DIRECT) {
-            ucxCtx.rxReqs[idx] = UcxPostRxReq(UCX_MSG_TAG_DIRECT, ucxCtx.eagerSize, NULL);
-            ucxCtx.rxReqs[idx]->idx = idx;
-            req = ucxCtx.rxReqs[idx];
-        } else {
-            return NULL;
-        }
     } else {
         req->msgBuf = buf;
     }
 
     return req;
 }
 
+static inline UcxRequest* UcxPostRxReq(ucp_tag_t tag, size_t size,
+                                       ucp_tag_message_h msg)
+{
+    UcxRequest *req = UcxPostRxReqInternal(tag, size, msg);
+    int idx = req->idx;
+
+    do {
+        if (req->completed) {
+            UCX_REQUEST_FREE(req);
+
+            if (tag & UCX_MSG_TAG_EAGER) {
+                req = UcxPostRxReqInternal(UCX_MSG_TAG_EAGER, ucxCtx.eagerSize, NULL);
+                ucxCtx.rxReqs[idx] = req;
+                std::swap(req->idx, idx);
```

Comment:
swap seems to be excessive, because you are working with the same reqs element (so `idx` will remain the same)
can do just `req->idx = idx;`

Reply:
I see. Is it guaranteed that every req will have the same idx? I wasn't sure if I could assume that here, so I used the swap to preserve idx assignment relationship from the recursive version of the function.

Reply:
here you are posting new request to the specific index of the requests array. If  it is completed immediately, you try to repost it again, but the index remains the same.